Distribution/Server/Features/ReverseDependencies.hs

```diff
@@ -77,69 +85,119 @@ data ReverseResource = ReverseResource {
+    rendRevList :: [ReverseRender],
+    rendFilterCount :: (Int, Int),
+    rendPageTotal :: Int
+}
+
+-- data Node = Node {id::Int, label::String} deriving Generic
+data Edge = Edge {
+    id::Int,
+    name::String,
+    deps::[String]
+    } deriving Generic
+-- data JGraph = JGraph { nodes::[Node], edges::[Edge]} deriving Generic
+-- instance ToJSON Node
+instance ToJSON Edge
+-- instance ToJSON JGraph
+-- instance ToJSON PackageName
+
 reverseFeature :: CoreFeature
-               -> Chan (IO (Map PackageName [Version]))
-               -> Hook (Map PackageName [Version] -> IO ())
-               -> Cache.Cache [(PackageName, Int, Int)]
+               -> Chan (IO [PackageId])
+               -> VersionsFeature
+               -> StateComponent AcidState ReverseIndex
+               -> MemState ReverseIndex
+               -> Hook [PackageId] ()
                -> ReverseFeature
 
 reverseFeature CoreFeature{..}
-               reverseStream reverseUpdateHook reverseTopCache
+               reverseStream
+               VersionsFeature{..}
+               reverseState
+               calculatedRI
+               reverseHook
   = ReverseFeature{..}
   where
     reverseFeatureInterface = (emptyHackageFeature "reverse") {
-        featureResources = map ($reverseResource) []
-      , featurePostInit = forkIO transferReverse >> return ()
-      , featureDumpRestore = Just (return [], restoreBackup, testRoundtripByQuery (query GetReverseIndex))
+        featureResources = map ($ reverseResource) []
+      , featurePostInit = initReverseIndex
+      , featureState    = [abstractAcidStateComponent reverseState]
+      , featureCaches   = [
+            CacheComponent {
+              cacheDesc       = "reverse index",
+              getCacheMemSize = memSize <$> readMemState calculatedRI
+            }
+          ]
       }
 
-    transferReverse = forever $ do
-            revFunc <- readChan reverseStream
-            modded  <- revFunc
-            runHook' reverseUpdateHook modded
-
-    --TODO: this isn't a restore!
-    --      do we need a post init/restore hook for initialising caches?
-    restoreBackup = RestoreBackup
-          { restoreEntry    = \_ -> return $ Right restoreBackup
-          , restoreFinalize = return $ Right restoreBackup
-          , restoreComplete = do
-                putStrLn "Calculating reverse dependencies"
-                index <- fmap packageList $ query GetPackagesState
-                let revs = constructReverseIndex index
-                update $ ReplaceReverseIndex revs
-          }
+    initReverseIndex :: IO ()
+    initReverseIndex = do
+            index <- queryGetPackageIndex
+            let ReverseIndex _ revdeps nodemap = constructReverseIndex index
+                assoc = takeWhile (\(a,_) -> a < Bimap.size nodemap) $ Arr.assocs . Gr.transposeG $ revdeps
+            forM_ (map (\(a,b) -> (nodemap Bimap.!> a, map (nodemap Bimap.!>) b)) assoc) $ uncurry setReverse
```

Comment:
This is almost certainly quadratic since it's doing each package one by one rather than in bulk.